In [114]:
import pandas as pd
import numpy as np

import matplotlib

import matplotlib.pyplot as plt

from pandas.tools.plotting import scatter_matrix

matplotlib.style.use('ggplot')
%matplotlib inline  


### Load Training Data:

In [115]:
#dataset column names:

col_names = ['id','cycle','setting1','setting2','setting3','s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']

# id: engine id
# cycle: cycle number. a sequence starts from 1 to the cycle # where failure has happened
# setting?: engine setting
# s1 to s21: sensors measurments on each cycle
# the dataset contains data of 100 engines with each engine run for number cycles to failure

In [116]:
#load training data

df_train_raw = pd.read_csv('data/PM_train.txt', sep = ' ', header=None)
df_train_raw.head()

0   1       2       3      4       5       6        7        8      9  ...  \
0   1   1 -0.0007 -0.0004  100.0  518.67  641.82  1589.70  1400.60  14.62 ...   
1   1   2  0.0019 -0.0003  100.0  518.67  642.15  1591.82  1403.14  14.62 ...   
2   1   3 -0.0043  0.0003  100.0  518.67  642.35  1587.99  1404.20  14.62 ...   
3   1   4  0.0007  0.0000  100.0  518.67  642.35  1582.79  1401.87  14.62 ...   
4   1   5 -0.0019 -0.0002  100.0  518.67  642.37  1582.85  1406.22  14.62 ...   

        18      19    20   21    22     23     24       25  26  27  
0  8138.62  8.4195  0.03  392  2388  100.0  39.06  23.4190 NaN NaN  
1  8131.49  8.4318  0.03  392  2388  100.0  39.00  23.4236 NaN NaN  
2  8133.23  8.4178  0.03  390  2388  100.0  38.95  23.3442 NaN NaN  
3  8133.83  8.3682  0.03  392  2388  100.0  38.88  23.3739 NaN NaN  
4  8133.80  8.4294  0.03  393  2388  100.0  38.90  23.4044 NaN NaN  

[5 rows x 28 columns]

In [117]:
#drop extra space columnn
df_train_raw.drop([26,27], axis=1, inplace='True')

In [118]:
#assign column names
df_train_raw.columns = col_names
df_train_raw.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5   ...        s12      s13      s14     s15   s16  s17   s18    s19  \
0  14.62   ...     521.66  2388.02  8138.62  8.4195  0.03  392  2388  100.0   
1  14.62   ...     522.28  2388.07  8131.49  8.4318  0.03  392  2388  100.0   
2  14.62   ...     522.42  2388.03  8133.23  8.4178  0.03  390  2388  100.0   
3  14.62   ...     522.86  2388.08  8133.83  8.3682  0.03  392  2388  100.0   
4  14.62   ...     522.19  2388.04  8133.80  8.4294  0.03  393  2388  100.0   

     s20      s21  
0  39.06  23.4190  
1  39.00  23.4236  
2  38.95  23.3442  
3  38.88  23.3739  
4  38.90  23.4044  

[5 rows x 26 columns]

In [119]:
df_train_raw.describe()

id         cycle      setting1      setting2  setting3  \
count  20631.000000  20631.000000  20631.000000  20631.000000   20631.0   
mean      51.506568    108.807862     -0.000009      0.000002     100.0   
std       29.227633     68.880990      0.002187      0.000293       0.0   
min        1.000000      1.000000     -0.008700     -0.000600     100.0   
25%       26.000000     52.000000     -0.001500     -0.000200     100.0   
50%       52.000000    104.000000      0.000000      0.000000     100.0   
75%       77.000000    156.000000      0.001500      0.000300     100.0   
max      100.000000    362.000000      0.008700      0.000600     100.0   

                 s1            s2            s3            s4            s5  \
count  2.063100e+04  20631.000000  20631.000000  20631.000000  2.063100e+04   
mean   5.186700e+02    642.680934   1590.523119   1408.933782  1.462000e+01   
std    6.537152e-11      0.500053      6.131150      9.000605  3.394700e-12   
min    5.186700e+02    641.210000   1571.040000   1382.250000  1.462000e+01   
25%    5.186700e+02    642.325000   1586.260000   1402.360000  1.462000e+01   
50%    5.186700e+02    642.640000   1590.100000   1408.040000  1.462000e+01   
75%    5.186700e+02    643.000000   1594.380000   1414.555000  1.462000e+01   
max    5.186700e+02    644.530000   1616.910000   1441.490000  1.462000e+01   

           ...                s12           s13           s14           s15  \
count      ...       20631.000000  20631.000000  20631.000000  20631.000000   
mean       ...         521.413470   2388.096152   8143.752722      8.442146   
std        ...           0.737553      0.071919     19.076176      0.037505   
min        ...         518.690000   2387.880000   8099.940000      8.324900   
25%        ...         520.960000   2388.040000   8133.245000      8.414900   
50%        ...         521.480000   2388.090000   8140.540000      8.438900   
75%        ...         521.950000   2388.140000   8148.310000      8.465600   
max        ...         523.380000   2388.560000   8293.720000      8.584800   

                s16           s17      s18      s19           s20  \
count  2.063100e+04  20631.000000  20631.0  20631.0  20631.000000   
mean   3.000000e-02    393.210654   2388.0    100.0     38.816271   
std    1.556432e-14      1.548763      0.0      0.0      0.180746   
min    3.000000e-02    388.000000   2388.0    100.0     38.140000   
25%    3.000000e-02    392.000000   2388.0    100.0     38.700000   
50%    3.000000e-02    393.000000   2388.0    100.0     38.830000   
75%    3.000000e-02    394.000000   2388.0    100.0     38.950000   
max    3.000000e-02    400.000000   2388.0    100.0     39.430000   

                s21  
count  20631.000000  
mean      23.289705  
std        0.108251  
min       22.894200  
25%       23.221800  
50%       23.297900  
75%       23.366800  
max       23.618400  

[8 rows x 26 columns]

In [120]:

def prepare_train_data (df_in, period):
    
    """
    df_in: input dataframe
    period: number of cycles used in classification modelling
    returns the input dataframe with additional 3 label columns: ttf, label_bnc, and label_mcc
    
    add labels to the training data:
        for regression: ttf (time-to-failure) = each cycle# for an engine subtracted from the last cycle# of the same engine
        for binary classification: label_bnc = if ttf is <= parameter period then 1 else 0 (values = 0,1)
        for multi-class classification: label_mcc = 2 if ttf <= 0.5*period paramter, 1 if ttf<= period, else 2 (values = 0,1,2)
    """
    
    #create regression label
    
    #make a dataframe to hold the last cycle for each enginge in the dataset
    df_max_cycle = pd.DataFrame(df_in.groupby('id')['cycle'].max())
    df_max_cycle.reset_index(level=0, inplace=True)
    df_max_cycle.columns = ['id', 'last_cycle']
    
    #add time-to-failure ttf as a new column - regression label
    df_in = pd.merge(df_in, df_max_cycle, on='id')
    df_in['ttf'] = df_in['last_cycle'] - df_in['cycle']
    df_in.drop(['last_cycle'], axis=1, inplace='True')
    
    #create binary classification label
    df_in['label_bnc'] = df_in['ttf'].apply(lambda x: 1 if x <= period else 0)
    
    #create multi-class classification label
    df_in['label_mcc'] = df_in['ttf'].apply(lambda x: 2 if x <= period/2 else 1 if x <= period else 0)
    
    return df_in
    

In [121]:
#add labels to training dataframe

df_train = prepare_train_data (df_train_raw, 30)
df_train.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5    ...         s15   s16  s17   s18    s19    s20      s21  ttf  \
0  14.62    ...      8.4195  0.03  392  2388  100.0  39.06  23.4190  191   
1  14.62    ...      8.4318  0.03  392  2388  100.0  39.00  23.4236  190   
2  14.62    ...      8.4178  0.03  390  2388  100.0  38.95  23.3442  189   
3  14.62    ...      8.3682  0.03  392  2388  100.0  38.88  23.3739  188   
4  14.62    ...      8.4294  0.03  393  2388  100.0  38.90  23.4044  187   

   label_bnc  label_mcc  
0          0          0  
1          0          0  
2          0          0  
3          0          0  
4          0          0  

[5 rows x 29 columns]

### Load Training Data:

In [122]:
#load testing data

df_test_raw = pd.read_csv('data/PM_test.txt', sep = ' ', header=None)
df_test_raw.head()


0   1       2       3      4       5       6        7        8      9  ...  \
0   1   1  0.0023  0.0003  100.0  518.67  643.02  1585.29  1398.21  14.62 ...   
1   1   2 -0.0027 -0.0003  100.0  518.67  641.71  1588.45  1395.42  14.62 ...   
2   1   3  0.0003  0.0001  100.0  518.67  642.46  1586.94  1401.34  14.62 ...   
3   1   4  0.0042  0.0000  100.0  518.67  642.44  1584.12  1406.42  14.62 ...   
4   1   5  0.0014  0.0000  100.0  518.67  642.51  1587.19  1401.92  14.62 ...   

        18      19    20   21    22     23     24       25  26  27  
0  8125.55  8.4052  0.03  392  2388  100.0  38.86  23.3735 NaN NaN  
1  8139.62  8.3803  0.03  393  2388  100.0  39.02  23.3916 NaN NaN  
2  8130.10  8.4441  0.03  393  2388  100.0  39.08  23.4166 NaN NaN  
3  8132.90  8.3917  0.03  391  2388  100.0  39.00  23.3737 NaN NaN  
4  8129.54  8.4031  0.03  390  2388  100.0  38.99  23.4130 NaN NaN  

[5 rows x 28 columns]

In [123]:
#drop extra space columnn
df_test_raw.drop([26,27], axis=1, inplace='True')

#assign column names
df_test_raw.columns = col_names
df_test_raw.head()


id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1    0.0023    0.0003     100.0  518.67  643.02  1585.29  1398.21   
1   1      2   -0.0027   -0.0003     100.0  518.67  641.71  1588.45  1395.42   
2   1      3    0.0003    0.0001     100.0  518.67  642.46  1586.94  1401.34   
3   1      4    0.0042    0.0000     100.0  518.67  642.44  1584.12  1406.42   
4   1      5    0.0014    0.0000     100.0  518.67  642.51  1587.19  1401.92   

      s5   ...        s12      s13      s14     s15   s16  s17   s18    s19  \
0  14.62   ...     521.72  2388.03  8125.55  8.4052  0.03  392  2388  100.0   
1  14.62   ...     522.16  2388.06  8139.62  8.3803  0.03  393  2388  100.0   
2  14.62   ...     521.97  2388.03  8130.10  8.4441  0.03  393  2388  100.0   
3  14.62   ...     521.38  2388.05  8132.90  8.3917  0.03  391  2388  100.0   
4  14.62   ...     522.15  2388.03  8129.54  8.4031  0.03  390  2388  100.0   

     s20      s21  
0  38.86  23.3735  
1  39.02  23.3916  
2  39.08  23.4166  
3  39.00  23.3737  
4  38.99  23.4130  

[5 rows x 26 columns]

In [124]:
# Load the truth data - actual 'ttf' for test data

df_truth = pd.read_csv('data/PM_truth.txt', sep = ' ', header=None)
df_truth.head()

0   1
0  112 NaN
1   98 NaN
2   69 NaN
3   82 NaN
4   91 NaN

In [125]:
#drop extra empty column and rename remaining 'ttf'

df_truth.drop([1], axis=1, inplace='True')
df_truth.columns = ['ttf']
df_truth.head()

ttf
0  112
1   98
2   69
3   82
4   91

In [126]:
def prepare_test_data(df_test_in, df_truth_in, period):
    
    """
    extract the last cycle for each enginge id to get the last record for each engine
    add the truth dataframe column to test dataframe as lable 'ttf'
    add other classification labels
    
    """
    
    df_tst_last_cycle = pd.DataFrame(df_test_in.groupby('id')['cycle'].max())
    
    df_tst_last_cycle.reset_index(level=0, inplace=True)
    df_tst_last_cycle.columns = ['id', 'last_cycle']
    
    df_test_in = pd.merge(df_test_in, df_tst_last_cycle, on='id')


    df_test_in = df_test_in[df_test_in['cycle'] == df_test_in['last_cycle']]

    df_test_in.drop(['last_cycle'], axis=1, inplace='True')
    
    df_test_in.reset_index(drop=True, inplace=True)
    
    df_test_in = pd.concat([df_test_in, df_truth], axis=1)
    
    #create binary classification label
    df_test_in['label_bnc'] = df_test_in['ttf'].apply(lambda x: 1 if x <= period else 0)
    
    #create multi-class classification label
    df_test_in['label_mcc'] = df_test_in['ttf'].apply(lambda x: 2 if x <= period/2 else 1 if x <= period else 0)

    return df_test_in

In [127]:
df_test = prepare_test_data(df_test_raw, df_truth, 30)
df_test.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1     31   -0.0006    0.0004     100.0  518.67  642.58  1581.22  1398.91   
1   2     49    0.0018   -0.0001     100.0  518.67  642.55  1586.59  1410.83   
2   3    126   -0.0016    0.0004     100.0  518.67  642.88  1589.75  1418.89   
3   4    106    0.0012    0.0004     100.0  518.67  642.78  1594.53  1406.88   
4   5     98   -0.0013   -0.0004     100.0  518.67  642.27  1589.94  1419.36   

      s5    ...         s15   s16  s17   s18    s19    s20      s21  ttf  \
0  14.62    ...      8.4024  0.03  393  2388  100.0  38.81  23.3552  112   
1  14.62    ...      8.4505  0.03  391  2388  100.0  38.81  23.2618   98   
2  14.62    ...      8.4119  0.03  395  2388  100.0  38.93  23.2740   69   
3  14.62    ...      8.4634  0.03  395  2388  100.0  38.58  23.2581   82   
4  14.62    ...      8.4362  0.03  394  2388  100.0  38.75  23.4117   91   

   label_bnc  label_mcc  
0          0          0  
1          0          0  
2          0          0  
3          0          0  
4          0          0  

[5 rows x 29 columns]

### Regression Modelling before Feature Engineering & Selection:

In [128]:
#Prepare data for regression model

from sklearn import linear_model
from sklearn.ensemble import RandomForestRegressor

sensor_cols = ['s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']


X_train = np.array(df_train[sensor_cols])
y_train = np.array(df_train['ttf'])

X_test = np.array(df_test[sensor_cols])
y_test = np.array(df_test['ttf'])

In [129]:
#try linear regression

linreg = linear_model.LinearRegression()
linreg.fit(X_train, y_train)
linreg.score(X_test, y_test)



0.40887945783383206

In [130]:
#try LASSO

lasso = linear_model.Lasso()
lasso.fit(X_train, y_train)
lasso.score(X_test, y_test)

0.31404765101730103

In [131]:
#try ridge

rdg = linear_model.Ridge()
rdg.fit(X_train, y_train)
rdg.score(X_test, y_test)

0.40827440087526334

In [132]:
#try Random Forest

rf = RandomForestRegressor(n_estimators=20, max_depth=4)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.37886332486260066

### Feature Engineering:

In [133]:
#exclude enging id and cycle number from the input features:

features = ['setting1','setting2','setting3','s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']
sensor_cols = ['s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']

In [134]:
def add_features(df_in, rolling_win_size):
    
    """
    add rolling average and rolling standard deviation for specific rolling window size
    
    
    """
    
    sensor_av_cols = [nm.replace('s', 'av') for nm in sensor_cols]
    sensor_sd_cols = [nm.replace('s', 'sd') for nm in sensor_cols]
    
    df_out = pd.DataFrame()
    
    ws = rolling_win_size
    
    #calculate rolling stats for each engine id
    
    for m_id in pd.unique(df_in.id):
    
        # get a subset for each engine sensors
        df_engine = df_in[df_in['id'] == m_id]
        df_sub = df_engine[sensor_cols]

    
        # get rolling mean for the subset
        av = df_sub.rolling(ws, min_periods=1).mean()
        av.columns = sensor_av_cols
    
        # get the rolling standard deviation for the subset
        sd = df_sub.rolling(ws, min_periods=1).std().fillna(0)
        sd.columns = sensor_sd_cols
    
        # combine the two new subset dataframes columns to the engine subset
        new_ftrs = pd.concat([df_engine,av,sd], axis=1)
    
        # add the new features rows to the output dataframe
        df_out = pd.concat([df_out,new_ftrs])
        
    return df_out


In [135]:
#create training dataframe that include new features with window size = 5

df_train_fe = add_features(df_train_raw, 5)
df_train_fe.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5    ...         sd12      sd13      sd14      sd15  sd16      sd17  \
0  14.62    ...     0.000000  0.000000  0.000000  0.000000   0.0  0.000000   
1  14.62    ...     0.438406  0.035355  5.041671  0.008697   0.0  0.000000   
2  14.62    ...     0.404475  0.026458  3.717450  0.007640   0.0  1.154701   
3  14.62    ...     0.495950  0.029439  3.050906  0.028117   0.0  1.000000   
4  14.62    ...     0.432574  0.025884  2.651326  0.025953   0.0  1.095445   

   sd18  sd19      sd20      sd21  
0   0.0   0.0  0.000000  0.000000  
1   0.0   0.0  0.042426  0.003253  
2   0.0   0.0  0.055076  0.044573  
3   0.0   0.0  0.076322  0.037977  
4   0.0   0.0  0.073621  0.033498  

[5 rows x 68 columns]

In [136]:
#add labels to the training dataframe

df_train_fin = prepare_train_data (df_train_fe, 30)

In [137]:
#create test dataframe that include new features with window size = 5

df_test_fe = add_features(df_test_raw, 5)

In [138]:
#add labels to the test dataframe

df_test_fin = prepare_test_data (df_test_fe, df_truth, 30)

### Regression Modelling after Feature Engineering:

In [139]:
# input features

all_cols = df_train_fin.columns.tolist()

#exclude labels, id, and cycle

exc_cols = ['id', 'cycle', 'ttf', 'label_bnc','label_mcc']
features = [i for i in all_cols if i not in exc_cols]

print(str(len(features)) + ' features:\n')
print(features)

66 features:

['setting1', 'setting2', 'setting3', 's1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11', 's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20', 's21', 'av1', 'av2', 'av3', 'av4', 'av5', 'av6', 'av7', 'av8', 'av9', 'av10', 'av11', 'av12', 'av13', 'av14', 'av15', 'av16', 'av17', 'av18', 'av19', 'av20', 'av21', 'sd1', 'sd2', 'sd3', 'sd4', 'sd5', 'sd6', 'sd7', 'sd8', 'sd9', 'sd10', 'sd11', 'sd12', 'sd13', 'sd14', 'sd15', 'sd16', 'sd17', 'sd18', 'sd19', 'sd20', 'sd21']


In [140]:
#try some feature selection

from sklearn.feature_selection import SelectFromModel
from sklearn.linear_model import LassoCV


X_train, y_train = np.array(df_train_fin[features]), np.array(df_train_fin['ttf'])
X_test, y_test = np.array(df_test_fin[features]), np.array(df_test_fin['ttf'])
                                             
clf = LassoCV()
                                             
sfm = SelectFromModel(clf, threshold=0.1)

sfm.fit(X_train,y_train)

X_train_transform = sfm.transform(X_train)
X_test_transform = sfm.transform(X_test)

print(X_train_transform.shape)

lasso = linear_model.Lasso(alpha=0.1)
lasso.fit(X_train_transform, y_train)
lasso.score(X_test_transform, y_test)



(20631, 17)


0.33554349759895419

In [141]:
#try standardize the data

from sklearn.preprocessing import StandardScaler

sc = StandardScaler()

X_train = np.array(df_train_fin[features])
y_train = np.array(df_train_fin['ttf'])

X_test = np.array(df_test_fin[features])
y_test = np.array(df_test_fin['ttf'])

#standardized features

X_train_sc = sc.fit_transform(X_train)
y_train_sc = sc.fit_transform(y_train)

X_test_sc = sc.fit_transform(X_test)
y_test_sc = sc.fit_transform(y_test)

D:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:429: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, _DataConversionWarning)
D:\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:586: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  warnings.warn(DEPRECATION_MSG_1D, DeprecationWarning)
D:\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:649: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  warnings.warn(DEPRECATION_MSG_1D, DeprecationWarning)
D:\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:586: DeprecationW

In [142]:
#try Linear Regression after feature engineering

linreg = linear_model.LinearRegression()
linreg.fit(X_train, y_train)
linreg.score(X_test, y_test)

0.34749933637137076

In [143]:
#try lASSO after feature engineering

lasso = linear_model.Lasso()
lasso.fit(X_train, y_train)
lasso.score(X_test, y_test)

0.31973657958920954

In [144]:
#try Ridge Regression after feature engineering

rdg = linear_model.Ridge()
rdg.fit(X_train_sc, y_train_sc)
rdg.score(X_test_sc, y_test_sc)

0.72042352828539458

In [145]:
#try Random Forest after feature engineering

rf = RandomForestRegressor(n_estimators=20, max_depth=4)
rf.fit(X_train_sc, y_train_sc)
rf.score(X_test_sc, y_test_sc)

0.77692196547666892

### Notes:  

__R^2__ after feature engineering  / before feature engineering

    Linear Regression         0.409 / 0.409  
    LASSO                     0.314 / 0.314  
    Ridge Regrassion          0.720 / 0.408  
    Random Forest             0.778 / 0.377  

Linear Regression and LASSO models scored very low both before and after feature engineering. This mainly due to multi collinearity between the features as shown in the correlation matrix below.   

Ridge Regression model which is robust to multi-collinearity has improved score after feature engineering.  

Random Forest, a non-linear model has also shown improvement after added features.  



In [146]:
feature_cor = df_train_fin[features].corr()

In [147]:
print(feature_cor[abs(feature_cor) > 0.7])

          setting1  setting2  setting3   s1        s2        s3        s4  \
setting1       1.0       NaN       NaN  NaN       NaN       NaN       NaN   
setting2       NaN       1.0       NaN  NaN       NaN       NaN       NaN   
setting3       NaN       NaN       NaN  NaN       NaN       NaN       NaN   
s1             NaN       NaN       NaN  1.0       NaN       NaN       NaN   
s2             NaN       NaN       NaN  NaN  1.000000       NaN  0.714949   
s3             NaN       NaN       NaN  NaN       NaN  1.000000       NaN   
s4             NaN       NaN       NaN  NaN  0.714949       NaN  1.000000   
s5             NaN       NaN       NaN  1.0       NaN       NaN       NaN   
s6             NaN       NaN       NaN  NaN       NaN       NaN       NaN   
s7             NaN       NaN       NaN  NaN -0.702136       NaN -0.793130   
s8             NaN       NaN       NaN  NaN       NaN       NaN  0.746852   
s9             NaN       NaN       NaN  NaN       NaN       NaN       NaN   